src/foam/dao/AbstractDAO.java

```diff
@@ -21,25 +21,25 @@ public DAO orderBy(Comparator comparator) {
     return new OrderedDAO().setOrder(comparator).setDelegate(this);
   }
 
-  public DAO skip(int count) {
+  public DAO skip(long count) {
     return new SkipDAO().setSkip(count).setDelegate(this);
   }
 
-  public DAO limit(int count) {
+  public DAO limit(long count) {
     return new LimitedDAO().setLimit(count).setDelegate(this);
   }
 
   public void pipe_(X x, foam.dao.Sink sink) {
     throw new UnsupportedOperationException();
   }
 
-  protected Sink decorateSink_(Sink sink, Long skip, Long limit, Comparator order, Predicate predicate) {
-    if ( limit != null ) {
-      sink = new LimitedSink().setLimit(limit.intValue()).setDelegate(sink);
+  protected Sink decorateSink_(Sink sink, long skip, long limit, Comparator order, Predicate predicate) {
+    if ( limit >= 0 ) {
```

Comment:
This should be:
if ( limit < 9007199254740991 ) {
Or, add MAX_SAFE_INTEGER as a constant and set it to that value.